### ISSUE TITLE
[FR]Delivery note

### ISSUE THREAD

**Issue by martonmiklos** [inventree/InvenTree#7023]
### Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find a similar feature request

### Problem statement

Sometimes when moving stuff between warehouse locations it is needed to have some sort of paper about the movement (security check etc.) It would be great to have a report type which could be printed out.

### Suggested solution

If the feature on the #7001  implemented then it would be straightforward to add this feature. 
On the other hand there could be a "view/wizard" where source and destination could be selected to filter out the stock movements for the report.

### Describe alternatives you've considered

-

### Examples of other systems

-

### Do you want to develop this?

- [X] I want to develop this.

**Comment by SchrodingersGat** [comment]
@martonmiklos adding the option within a generic "transfer stock" dialog might be enough, but it might require a new "stock transfer" model which can be built up by the user and then enacted once all stock items are selected for transfer. Maybe a "Transfer Order" which would be similar to a Sales Order?

**Comment by martonmiklos** [comment]
I would create a separate model for this purpose: later on we can add more features like incoming approval, etc. using that.

**Comment by SchrodingersGat** [comment]
Still relevent but needs someone to either fund or implement

**Comment by jacobfelknor** [comment]
> [@martonmiklos](https://github.com/martonmiklos) adding the option within a generic "transfer stock" dialog might be enough, but it might require a new "stock transfer" model which can be built up by the user and then enacted once all stock items are selected for transfer. Maybe a "Transfer Order" which would be similar to a Sales Order?

@SchrodingersGat I'm not promising anything, but this is something I'm interested in for my instance too. I'm going to start looking at it. If I make any good progress on it I will open source it.

We are primarily targeting stock transfer requests from our technicians. In our scenario, we'd like the option for this Transfer Order to have two modes.... 1) transfer all queued items to a specified location (most common).... and 2) _Consume_ all queued items so that they are removed from inventory (less common, but it happens).

Scenario 2 is effectively an ad-hoc Build Order where we don't care to keep in inventory the thing being produced. We get around this currently by manually removing stock for these ad-hoc cases, but having the "Transfer Order" is a good way of documenting what happened in a consolidated location.

Would you prefer this be a plugin or a native feature? Happy to bounce ideas on this

**Comment by SchrodingersGat** [comment]
I think that there has been enough interest in this that it makes sense as a first-class procedure - as long as it fits well with the existing workflows

We must retain the existing stock transaction options (i.e. manual immediate operations without tracking) - I do not want to go down the road of enforcing a stock transfer request for each and every operation that technicians must perform.

However, having the option of a more "documented" process would be good.

Some thoughts:

If the stock items are "deleted" after the operation is complete, they will disappear from the records in the stock transfer operation. Depending on the type of traceability you need, this may be undesirable

Let's discuss what your ideas are around the implementation of such a feature

**Comment by jacobfelknor** [comment]
Agree completely that we should retain the ability to do manual stock adjustments at any time.

I will think more about the case where stock is "deleted" or "consumed" by the operation. I agree that if it disappears from the records this isn't what we want. Ideally, from the user perspective it will behave similarly to the "consumed stock" panel of the build order.

When I have a more concrete idea of how I'd approach it, I will let you know.

**Comment by Petrox** [comment]
Just a quick idea for thought:

What if the "Delivery notes" would be treated like some specialized stock
locations?

They could behave like a box or shelf, you can add remove things from them,
and they wouldn't technically disappear from the stock of the company just
when you prepare them or maybe move them between sites, but they would be
on stock "in transit from A to B"

We already planned to use "travel boxes" for the same concept, it was just
too cumbersome.

When it arrives to the recipient different actions could be run from moving
items to where they really are on the other end, changing ownerships
(SalesOrder), further processed in manufacturing, etc.

The only thing this does not cover is the proper tracing so "what was on
DeliverNote 202 last week", and also I don't think delivery notes should be
dynamic in their content after some kind of "locked" state.

Maybe I'm on a wrong track but a delivery note is not very far from
something like a shipment tracking and packaging logically. Maybe these two
features (birds) could be implemented in one step (stone). To align the
deliverynote with the DHL/Fedex/etc states and tracking numbers and apis
via plugins...

**Comment by matmair** [comment]
Please do not feature creep this into shipment tracking

**Comment by jacobfelknor** [comment]
It's looking unlikely I'll get any significant work done on this before the holiday break, but I have not forgotten about it 😁

**Comment by jacobfelknor** [comment]
This is very much a work in progress, but I've made some decent headway on setting up a `TransferOrder` at [jacobfelknor/InvenTree@transfer_order](https://github.com/jacobfelknor/InvenTree/tree/transfer_order).

I'm basing it off other existing orders, such as a `SalesOrder`, except as @matmair has expressed, there is no concept of a "shipment" and stock is allocated directly to a line item. I'm only to the point where the order details, line items, and un-tracked allocations can be made. I still wanted to check in with @SchrodingersGat and @matmair early before I get much further to address any concerns.

My idea is that upon completion of the order, all allocated stock is moved from its current location to the order's destination location using the existing `StockItem.move()` function.

In order to make reporting useful, there will need to be some information captured by the order upon completion independent of a direct relationship to the stock items themselves. This way, a person can reference the Transfer at any later time and see what happened, regardless of whether the affected stock items are still around. I'm not quite sure what this looks like yet, but some information stored in a `JSONField` directly on the `TransferOrder` itself may be sufficient (inspired by the `deltas` field on a stock tracking event).

And to reiterate and be clear, this is in no way replacing existing immediate stock transaction options - only optional additional "paperwork" to maintain records if you need it (like we do).

I plan to add settings to enable/disable the `TransferOrder` feature, reference patterns, permissions, etc. similar to other order types. Of course I'll also need to write tests/documentation for this to be considered complete.

**Comment by PaulGregor** [comment]
> This is very much a work in progress, but I've made some decent headway on setting up a `TransferOrder` at [jacobfelknor/InvenTree@transfer_order](https://github.com/jacobfelknor/InvenTree/tree/transfer_order).

Thank you so much, this feature is very needed.

**Comment by matmair** [comment]
Looks pretty good already @jacobfelknor!

Regarding tracking: Maybe we can use the basics from the stock tracking entry model for this by making them a bit more generic and having that as a general building block for new things?  
Having a more generalized, immutable event storage might be useful for other things like #9996 too.

Regarding reporting, my recommendation would be to add a report type for TOs and maybe a sane default template for that.

Once you feel it is ready, I would encourage you to create a drag PR, that makes commenting on and discussing specific aspects easier.  
We can also add a feature flag and merge for example only a MVP once that is clean and work out things like rich tracking events later.

**Comment by matmair** [comment]
@jacobfelknor is this something that would be polished enough for inclusion in the next [4ish weeks](https://github.com/inventree/InvenTree/discussions/11199) or should be slated this for 1.3?

**Comment by jacobfelknor** [comment]
Hoping yes. I had to move off of it this week, but want to get back to it. If I feel like I start to fall further behind I'll let you know

**Comment by jacobfelknor** [comment]
As of 8635bfb2c18566eda9d6c9a41ba64656b8f3b2d4 on [jacobfelknor/InvenTree@transfer_order](https://github.com/jacobfelknor/InvenTree/tree/transfer_order), the Transfer Order now supports allocating serialized stock and moves all allocated items to the destination on order complete. This action also appears in the tracking history for the affected stock.

Still no reporting or settings, and there is quite a bit of polishing to do. But good progress tonight

**Comment by mabroens** [comment]
Hi All, is there a way to test the implementation? and or any idea when reporting would be availeble ?
this to me is a vital fucntion

Br Martijn

**Comment by jacobfelknor** [comment]
@mabroens, feel free to check out my branch at [jacobfelknor/InvenTree@transfer_order](https://github.com/jacobfelknor/InvenTree/tree/transfer_order) which includes a basic example report similar to other example order reports. Of course you can customize your report as you see fit.

It still relies on the transferred stock still existing, so any report should be generated before the stock is further manipulated

**Comment by SchrodingersGat** [comment]
@jacobfelknor can you please submit a PR for easy review? It's OK if it's not fully polished yet

**Comment by jacobfelknor** [comment]
Yes, opened at https://github.com/inventree/InvenTree/pull/11281

I see you've already switched this to the 1.3 milestone. I agree, was hoping to spend more time with it but you know how it goes sometimes.

**Comment by SchrodingersGat** [comment]
Thanks @jacobfelknor 👍 this gives us more time to assess / review / etc

**Comment by SchrodingersGat** [comment]
Closed in https://github.com/inventree/InvenTree/pull/11281
